### PR TITLE
Event queue

### PR DESCRIPTION
**Pull request by paulidale**
Partial implementation of an event queue on top of #18274 
This should be enough to get a feel for the design

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: ssl/priority_queue.c [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+static ossl_inline void pqueue_move_elem(OSSL_PQUEUE *pq, size_t from, size_t to)
+{
+    struct pq_heap_s *h = pq->heap;
+    struct pq_elem_s *e = pq->elements;
+
+    ASSERT_USED(pq, from);
+
+    h[to] = h[from];
+    e[h[to].index].posn = to;
+}
```
Comment (tmshort) [inline ssl/priority_queue.c:128]: Should the `from` element be cleared/removed?

File: doc/internal/man3/OSSL_EVENT.pod [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+subscribed event type I<sub>.
+
+=head1 RETURN VALUES
+
+B<ossl_event_queue_new> returns a new timer queue or NULL on failure.
+
+B<ossl_event_new> returns a new event or NULL on failure.
+
+B<ossl_event_get_type>, B<ossl_event_get_priority> and
+B<ossl_event_get_when> return the correspond even field.
```
Comment (mattcaswell) [inline doc/internal/man3/OSSL_EVENT.pod:185]: typo: corresponding
typo: event

File: include/internal/event_queue.h [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+# define OSSL_INTERNAL_EVENT_QUEUE_H
+# pragma once
+
+# include "internal/priority_queue.h"
+# include "internal/time.h"
+
+/*
+ * Opaque type holding an event.
+ */
+typedef struct ossl_event_s OSSL_EVENT;
```
Comment (mattcaswell) [inline include/internal/event_queue.h:20]: Structures normally end in `_st`. Ditto for other similar instances below.

File: include/internal/event_queue.h [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+/*
+ * Public type representing an event queue, the underlying structure being
+ * opaque.
+ */
+typedef struct ossl_event_queue_s OSSL_EVENT_QUEUE;
+
+/*
+ * Public type representing a event queue entry.
+ * It is public so that it can be embedded into other structures,
```
Comment (mattcaswell) [inline include/internal/event_queue.h:42]: Welll...internally public!

File: include/internal/event_queue.h [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                    void *payload, size_t payload_size)
+{
+    event->type = type;
+    event->priority = priority;
+    event->when = when;
+    event->ctx = ctx;
+    event->payload = payload;
+    event->payload_size = payload_size;
+    /*event->ref = 0;*/
+    /*event->flag_dynamic = 0;*/
```
Comment (mattcaswell) [inline include/internal/event_queue.h:73]: Remove?

File: ssl/event_queue.c [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        }
+    }
+    return r;
+}
+
+void ossl_event_queue_free(OSSL_EVENT_QUEUE *queue)
+{
+    OSSL_EVENT *e;
+
+    /* This isn't remotely correct yet */
```
Comment (mattcaswell) [inline ssl/event_queue.c:62]: What remains to be done?

File: ssl/event_queue.c [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+OSSL_EVENT *ossl_event_new(uint32_t type, uint32_t priority,
+                           OSSL_TIME when, void *ctx,
+                           void *payload, size_t payload_size)
+{
+    OSSL_EVENT *e = OPENSSL_malloc(sizeof(*e));
+
+    if (e != NULL) {
+        ossl_event_set(e, type, priority, when, ctx, payload, payload_size);
+        e->flag_dynamic = 1;
```
Comment (mattcaswell) [inline ssl/event_queue.c:82]: This makes me wonder why we need non-dynamic events.

File: ssl/event_queue.c [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    return e;
+}
+
+void ossl_event_free(OSSL_EVENT *event)
+{
+    if (event != NULL && event->flag_dynamic)
+        OPENSSL_free(event);
+}
+
+int ossl_event_queue_add(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event){
```
Comment (mattcaswell) [inline ssl/event_queue.c:93]: Move `{` to next line

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Conspicuous by their absence are event queue tests

File: ssl/event_queue.c [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    return ossl_pqueue_OSSL_EVENT_push(queue->events, event, &event->ref);
```
Comment (mattcaswell) [inline ssl/event_queue.c:94]: So in this event queue design adding events to the queue does not also dispatch them. For timed events that haven't "fired" yet this is fine. What about "immediate" events? How do we handle those? Presumably they are represented internally with a "when" of "now" (or is that necessary....do we continually want to be querying the system time for this type of event?)

Comment (paulidale) [reply ssl/event_queue.c:94]: I'd represent them with a "when" of zero rather than now (which should be wrapped in a macro).

I see running them immediately as a trivial change, if we wanted to.  There are a few potential concerns with doing so:
1. Raising two events of different priority simultaneously could go awry if done in reverse order (more likely if different pieces of code raise them -- e.g. during an event call back sequence).
2. It removes control of the dispatching from the calling application (i.e. stuff happens without calling _tick).
3. It could introduce recursion (i.e. locking) problems (e.g. calling an application callback which raises an immediate event via some action it does).

I've also figured out a way to avoid them hitting the main event queue if the performance loss of doing so is discovered to be a problem

**Comment by paulidale** [comment]
I didn't do tests because there is no guarantee that this approach will be used.

It definitely isn't quite ready to go at this point.

File: ssl/priority_queue.c (hunk 1 above)
Comment (paulidale) [reply ssl/priority_queue.c:128]: It doesn't have to be.  This is only called from the pop function which deals with the from side.

File: include/internal/event_queue.h [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    return event->payload;
+}
+
+
+/*
+ * Dynamically create and free events
+ */
+OSSL_EVENT *ossl_event_new(uint32_t type, uint32_t priority,
+                           OSSL_TIME when, void *ctx,
+                           void *payload, size_t payload_size);
```
Comment (levitte) [inline include/internal/event_queue.h:111]: `priority` and `when` are scheduling parameters.  As such, I would rather have them passed to the queuing calls, i.e. `ossl_event_queue_add()`.

Where those parameters are *stored*, I don't really care.

**Review by levitte — CHANGES_REQUESTED** [review]

File: ssl/event_queue.c (hunk 6 above)
Comment (paulidale) [reply ssl/event_queue.c:62]: The possibility of an optional callback for the user to clean things up is worth considering.
"remotely correct" wasn't a good choice of words.

File: include/internal/event_queue.h (hunk 10 above)
Comment (levitte) [reply include/internal/event_queue.h:111]: Er, I don't see `type` as a **scheduling** parameter.  Sure it's a dispatching parameter in so far that you choose subscribers to call based on the type, but I don't see that as **scheduling** per se

File: ssl/event_queue.c (hunk 7 above)
Comment (paulidale) [reply ssl/event_queue.c:82]: A non-dynamic event can be embedded into another structure.  i.e. cuts overhead allocating and freeing.
E.g. a re-transmit timer event will always exist with a connection -- do we need to allocate it separately?

Dropping non-dynamic is possible, but it feels like it will be useful.

File: include/internal/event_queue.h (hunk 10 above)
Comment (levitte) [reply include/internal/event_queue.h:111]: (see it like this: `priority` and `when` affects where the event ends up in the queue order.  `type` does not)

File: doc/internal/man3/OSSL_EVENT.pod [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+=item type
+
+A mandatory event type, which is a simple numeric identity, the
+meaning of which is not known by the event functionality itself.
+
+=item priority
+
+A manatory nonnegative integral quantity.  The lower the priority the earlier
+the event will be processes.
+
```
Comment (t8m) [inline doc/internal/man3/OSSL_EVENT.pod:80]: I liked the Richard's suggestion from our discussion that a type might have an inherent priority and we would not need a separate priority value.

File: doc/internal/man3/OSSL_EVENT.pod [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+reference to the destructor function are considered immutable, up until the
+event structure is destroyed.
+
+The reference to the auxilliary identifying material or to the payload,
+however, are considered mutable.  Any event handler may "steal" them and
+replace the reference to them in the event structure with NULL.  Stealing
+must be done with much care.
+
+Events may be embedded in another structure, as a static variable, or
+as an local variable in a function.  Events may also be dynamically allocated.
```
Comment (t8m) [inline doc/internal/man3/OSSL_EVENT.pod:112]: Honestly I cannot imagine how static or local variable events would work. Perhaps immediate events could work for static ones but local variables that can go out of scope any time? Especially if scheduling and dispatch is separate.

File: doc/internal/man3/OSSL_EVENT.pod [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+=over 4
+
+=item type
+
+A mandatory event type, which is a simple numeric identity, the
+meaning of which is not known by the event functionality itself.
+
+=item priority
+
+A manatory nonnegative integral quantity.  The lower the priority the earlier
```
Comment (t8m) [inline doc/internal/man3/OSSL_EVENT.pod:78]: Typo: mandatory

File: doc/internal/man3/OSSL_EVENT.pod (hunk 12 above)
Comment (t8m) [reply doc/internal/man3/OSSL_EVENT.pod:112]: Also, if we want them embeddable, the structure must be public (internally only of course), do then the ossl_event_set and/or the getters make any sense?

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
A few more comments:

I suppose the events will be sorted by `when` as the primary key and `priority` as the secondary key in the priority queue. In that case I wonder how dispatch will work if some high priority event has `when` in the past but there are multiple lower priority events which have `when` even further in the past. I suppose these will be dispatched first which is not necessarily correct.

My another concern is with the subscription concept - I wonder whether this isn't overcomplicating things, in a way that requires complicated setup of the subscriptions. Couldn't we have for example a simple static per-type callback table? I mean the callback data and the context should provide enough contextual data. Do we think we will have multiple subscribers for each event type?

File: include/internal/event_queue.h (hunk 10 above)
Comment (paulidale) [reply include/internal/event_queue.h:111]: I had a thought over night -> what about a new_and_add call?
The _set call disappears.

Both the new_and_add and the _add call take the full suite of parameters.
I expect that every call to _new will be immediately followed by a call to _add.

**Comment by paulidale** [comment]
@t8m, good points.

For dispatch point, it would be possible to extract a list of events that need to run now and sort them by priority rather than time.  The main list would be sorted by time only and immediate events would go into the second queue only.

I'm sure about the subscriptions either.  There are multiple options (in rough order of complexity):
- one event handler that dispatches by type -- this is kind of what @levitte did.
- a global array of static event handlers.
- a per SSL_METHOD array of static event handlers.
- installed per type handlers as per this.
- Event handler associated with the OSSL_EVENT object (caller sets what deals with the event).

They should all be fairly easy to convert between, e.g. the SSL_METHOD could install the per type handlers during initialisation or a single global event hander could dispatch by type to SSL_METHOD handlers.

File: doc/internal/man3/OSSL_EVENT.pod (hunk 12 above)
Comment (paulidale) [reply doc/internal/man3/OSSL_EVENT.pod:112]: The big question is do we want them to be embeddable or not?
It's simplier to not but more efficient to allow it.

We're already got way too many mallocs happening :crying_cat_face:

**Comment by paulidale** [comment]
Embeddable or not?

The reasons for making the event structure embeddable is efficiency.  I see two primary use cases:
- embedded in the caller's event context (the event lives with the event's data)
```c
struct {
    // my fields
    ...
    // event
    OSSL_EVENT event;
} mydata;

ossl_event_queue_add(queue, &mydata->event);
```
- embedded with the protocol structure (e.g. re-transmit events)
Plus a reduction in the number of malloc and free calls.

The reasons against are modularity and good coding practices.

Is this a premature optimisation?

**Comment by levitte** [comment]
>     * one event handler that dispatches by type -- this is kind of what @levitte did.

Uhm no.  My design allowed multiple subscribers per event type.  This was meant to allow a fork where needed...
for example, each stream could be handled by a separate subscriber, where the stream is identified by the indentifiers material attached to the event.  With all those subscribers associated with the same queue, I imagine that whatever code that parsed a frame and found the identifying material as well as the payload, would generate one READ event and add it to the queue, and the rest would be a dispatch.

**Comment by t8m** [comment]
> Embeddable or not?

I like the concept of embeddable events as I strongly suspect that for reasonable performance the library should avoid mallocs whenever possible. Retrofitting them into the design later when we'll find that the performance is sub-par might be quite a complicated effort.

IMO the biggest concern with embeddable events is the embedded event lifetime - i.e., how to ensure the enclosing object is not prematurely freed, or that the event is forcibly removed from the queue when freeing the enclosing object.

File: include/internal/event_queue.h (hunk 10 above)
Comment (levitte) [reply include/internal/event_queue.h:109]: > I had a thought over night -> what about a new_and_add call? The _set call disappears.

That implies that embedded event structures will no longer be supported, right?

> Both the new_and_add and the _add call take the full suite of parameters. I expect that every call to _new will be immediately followed by a call to _add.

That is, of course, a solution too...  nice escape from our conflict in opinion :wink:

Comment (paulidale) [reply include/internal/event_queue.h:109]: I was thinking that new_and_add() creates an event while add() uses an embedded event.

Comment (levitte) [reply include/internal/event_queue.h:109]: I'm a little confused...  how's the embedded event being populated?

File: ssl/time.c [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        FILETIME ft;
+    } now;
+
+    GetSystemTime(&st);
+    SystemTimeToFileTime(&st, &now.ft);
+    /* re-bias to 1/1/1970 */
+# ifdef  __MINGW32__
+    now.ul -= 116444736000000000ULL;
+# else
+    /* *INDENT-OFF* */
```
Comment (levitte) [inline ssl/time.c:29]: ??? Why is this necessary?

File: include/internal/event_queue.h (hunk 10 above)
Comment (paulidale) [reply include/internal/event_queue.h:109]: `_add(queue, &event, type, priority, when, ctx, payload, payload_size);`

File: ssl/time.c (hunk 14 above)
Comment (paulidale) [reply ssl/time.c:29]: No idea, this was just copied wholesale from `d1_lib.c`.  The duplicate we have in `crypto/bio/bss_dgram.c` doesn't include these directives.

Comment (levitte) [reply ssl/time.c]: There must have been a block of code that was formaterad in a special way for readability at some point in history...

The way this looks now, though, these indentation directives (that's what these comments are) do nothing of value. Remove them.

Comment (paulidale) [reply ssl/time.c]: Gone.

File: include/internal/event_queue.h (hunk 10 above)
Comment (levitte) [reply include/internal/event_queue.h:109]: I would call that `_set_and_add`, to go along with `_new_and_add`

File: include/internal/event_queue.h [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+/*
+ * Public type representing a event queue entry.
+ * It is (internally) public so that it can be embedded into other structures,
+ * it should otherwise be treated as opaque.
+ */
+struct ossl_event_st {
+    uint32_t type;              /* What type of event this is               */
+    uint32_t priority;          /* What priority this event has             */
+    OSSL_TIME when;             /* When the event is scheduled to happen    */
```
Comment (hlandau) [inline include/internal/event_queue.h:45]: I do think we can combine type and priority, and just call it type.

**Review by hlandau — CHANGES_REQUESTED** [review]
Looks good on the whole.

File: doc/internal/man3/OSSL_EVENT.pod [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+ OSSL_TIME ossl_event_time_until(OSSL_EVENT *event);
+ OSSL_TIME ossl_event_queue_time_until_next(const OSSL_EVENT_QUEUE *queue);
+
+ int ossl_event_queue_postpone(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event,
+                               OSSL_TIME how_long);
+ int ossl_event_queue_postpone_until(OSSL_EVENT_QUEUE *queue,
+                                     OSSL_EVENT *event,
+                                     OSSL_TIME when);
+
+ int ossl_event_queue_dispatch(OSSL_EVENT_QUEUE *queue);
```
[the remark this answers is not in the document]
Comment (paulidale) [reply doc/internal/man3/OSSL_EVENT.pod:55]: I like this suggestion, I'll update.

File: include/internal/event_queue.h (hunk 15 above)
Comment (paulidale) [reply include/internal/event_queue.h:38]: It's tempting to remove `type` altogether since it's not something the event queue actually cares about anymore.

**Comment by paulidale** [comment]
Ready for review

**Comment by mattcaswell** [comment]
We seem to have a couple of extra commits in this PR related to CI changes from #18639. They don't look like they belong here?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/internal/man3/OSSL_EVENT.pod [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+ void *ossl_event_get0_payload(const OSSL_EVENT *event, size_t *length);
+
+ int ossl_event_queue_schedule(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event);
+ int ossl_event_queue_delete(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event);
+
+ OSSL_TIME ossl_event_time_until(OSSL_EVENT *event);
+ OSSL_TIME ossl_event_queue_time_until_next(const OSSL_EVENT_QUEUE *queue);
+
+ int ossl_event_queue_postpone(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event,
+                               OSSL_TIME how_long);
```
[the remark this answers is not in the document]
Comment (paulidale) [reply doc/internal/man3/OSSL_EVENT.pod:51]: Yeah, dropped it.  Not particularly useful given the ACK manager design.

**Comment by paulidale** [comment]
Mix up history due to the merge mistake this morning.

**Review by mattcaswell — APPROVED** [review]

**Comment by paulidale** [comment]
Ping for second review

**Comment by paulidale** [comment]
Merged!
